modules/ardrivo/test/String.cxx

```diff
@@ -0,0 +1,215 @@
+	String test2 = "test";
+	String test3 = "est";
+	REQUIRE(test.indexOf(test2) == 0);
+	REQUIRE(test.indexOf(test3) == 1);
+}
+
+TEST_CASE("Modify in place a String removing chars from the provided index to the end of the String or from the provided index to index plus count" "[remove]")
+{
+	String test = "test";
+	test.remove(3);
+	REQUIRE_THAT(test.c_str(), Catch::Contains("tes"));
+}
+
+TEST_CASE("Check if string gets swapped" "[replace]")
+{
+	String test = "ab";
+	String test2 = "cd";
+	test.replace(test, test2);
+	REQUIRE_THAT(test.c_str(), Catch::EndsWith("cd"));
+}
+
+//Crap does not work without std::string
+TEST_CASE("Test if string is given a new capacity" "[reserve]")
+{ 
+	String cap = "";
+	REQUIRE(cap.length() == 0);
+
+	cap.reserve(25);
+	//REQUIRE(cap.length() == 63);
+}
+
+TEST_CASE("Sets a character of the String" "[setCharAt]")
+{
+	String test = "Hejsan";
+	test.setCharAt(1, '\x84');
+	REQUIRE_THAT(test.c_str(), Catch::EndsWith("H\x84jsan"));
+}
+
+TEST_CASE("Get a substring of a String" "[subsring]")
+{	 
+	String test = "test";
+	String newString = test.substring(2);
+	REQUIRE_THAT(newString.c_str(), Catch::Contains("st"));
+}
+
+TEST_CASE("Copies the String�s characters to the supplied buffer" "[toCharArray]")
+{
+	String test = "test";
+	char arr[5] = "init";
+	test.toCharArray(arr, test.length());
+	REQUIRE_THAT(arr, Catch::Contains("test"));
+}
+
+TEST_CASE("Converts a valid String to an integer. The input String should start with an integer number" "[toInt]") 
+{
+	String test = "1 and 23";
+	String test2 = "123";
+	String test3 = "-100";
+	String test4 = "--100";
+	String test5 = "+100";
+	auto num = test.toInt();
+	auto num2 = test2.toInt();
+	auto num3 = test3.toInt();
+	auto num4 = test4.toInt();
+	auto num5 = test5.toInt();
+	REQUIRE(num == 1);
+	REQUIRE(num2 == 123);
+	REQUIRE(num3 == 0);
+	REQUIRE(num4 == 0);
+	REQUIRE(num5 == 0);
+}
+
+TEST_CASE("Converts a valid String to an double. The input String should start with an double number" "[toDouble]")
+{
+	String test = "1.0323414535 and 23";
+	double num =  test.toDouble();
+	REQUIRE(num >= 1.03228);
+	REQUIRE(num <= 1.03240);
+}
+
+TEST_CASE("Converts a valid String to an float. The input String should start with an float number" "[toFloat]")
+{
+	String test = "33.34523345 stuff 23";
+	float num =  test.toFloat();		
+	REQUIRE(num >= 33.3445);
+	REQUIRE(num <= 33.3460);
+}
+
+TEST_CASE("Get a lower-case version of a String" "[toLowerCase]")
+{
+	String test = "I LIKE BANANAS";
+	test.toLowerCase();
+	REQUIRE_THAT(test.c_str(), Catch::Contains("i like bananas"));
+}
+
+TEST_CASE("Get a upper-case version of a String" "[toUpperCase]")
+{
+	String test = "i like bigger bananas";
+	test.toUpperCase();
+	REQUIRE_THAT(test.c_str(), Catch::Contains("I LIKE BIGGER BANANAS"));
+}
+
+TEST_CASE("Remove spaces in a string" "[trim]")
+{
+	String test = "i like bigger bananas";
+	test.trim();
+	REQUIRE_THAT(test.c_str(), Catch::Contains("ilikebiggerbananas"));
```

Comment:
Shoudnt trim only delete leading and trailing whitespaces? [arduino ref](https://www.arduino.cc/reference/en/language/variables/data-types/string/functions/trim)
Looking at our impl its deleting all whitespaces :thinking: 
```cpp
void String::trim() { m_u.erase(std::remove(m_u.begin(), m_u.end(), ' '), m_u.end()); }
```

Reply:
Correct test case would be:
```cpp
	String test = "    i like bigger bananas    ";
	test.trim();
	REQUIRE_THAT(test == "i like bigger bananas");
```

You have to write tests for whats supposed to be right not what actually works, otherwise you start testing if bugs work correctly